src/v/cloud_storage/remote_partition.cc

```diff
@@ -501,15 +502,18 @@ remote_partition::get_term_last_offset(model::term_id term) const {
 
 ss::future<std::vector<cluster::rm_stm::tx_range>>
 remote_partition::aborted_transactions(offset_range offsets) {
+    gate_guard guard(_gate);
     // Here we have to use kafka offsets to locate the segments and
     // redpanda offsets to extract aborted transactions metadata because
     // tx-manifests contains redpanda offsets.
     std::vector<cluster::rm_stm::tx_range> result;
-    auto first_it = _segments.upper_bound(offsets.begin);
-    if (first_it != _segments.begin()) {
+
+    // that's a stable btree iterator that makes key lookup on increment
+    auto first_it = upper_bound(offsets.begin);
+    if (first_it != begin()) {
         first_it = std::prev(first_it);
     }
-    for (auto it = first_it; it != _segments.end(); it++) {
+    for (auto it = first_it; it != end(); it++) {
```

Comment:
Based on the description of the problem seen, this still looks problematic. IIUC how the remote partition iterator works it's safe to increment it because it caches a key and performs a fresh lookup. However, imagine here that `it` is valid. Then a few lines down a reference to the value is taken:

```c++
        auto& st = it->second;
```

and then scheduling points occur that ostensibly are affecting the underlying container:

```
        auto tx = co_await ss::visit(
          st,
          [this, &st, offsets, offset_key = it->first](
```

Directly from the abseil documentation about invalidations:

> Importantly, insertions and deletions may invalidate outstanding iterators
pointers, and references to elements. Such invalidations are typically only
an issue if insertion and deletion operations are interleaved with the use of
more than one iterator, pointer, or reference simultaneously.

However, even if all that is ok here, in the special iterator in remote partition it still looks unsafe. Deference looks up the key (which will be present), and then unconditionally dereferences it. And I don't see any reason why it is guarnateed that the key actually exists int he container after scheduling point.

```
auto it = _map.get().find(*_key);
return *it;
```

Reply:
I think right now it exists because we never delete from `_segments`, we only `insert_or_assign()` to it. But I agree that generally we want better guarantees. 

cc @Lazin

Reply:
> But I agree that generally we want better guarantees.

I think Andrew's point is good. If interleaved modifications to the container are restricted to a set of operations that make this safe, then that should be documented, otherwise this isn't about better guarantees but rather about a latent bug.

I don't see any `erase` method (which is curious, how does it string? :)), but I do see an `insert_or_assign` where the value contains the type of things that look problematic if destructed (e.g. unique_ptrs etc...) while there is an open reference.

Reply:
Looking through the dereferences, it's not that straightforward, but nothing really jumped out as flagrantly incorrect, e.g. in the places we use the value state, I don't think we ever destruct the value. I think it's reasonably safe that callers that are able to dereference from the map expect the values to stay valid, and that dereferencers never destruct them, but it's definitely another thing to document since it seems fairly easy to misuse.

It's also worth thinking about what improvements we can bake into the APIs to make it harder to misuse, though nothing concrete comes to mind at the moment.

Reply:
> e.g. in the places we use the value state, I don't think we ever destruct the value

map::insert_or_assign will destruct the previous value, no?

Reply:
oh i see, the dereference looks up the value again

Reply:
I take that back, we do hold a refernece to the value here `        auto& st = it->second;` right?

Reply:
Yeah but the value references are safe to hold, provided that the underlying btree never erases anything and that other callers with references to the underlying values don't reset the value. Holding a reference to a `unique_ptr` should be fine on its own since it won't destruct if it leaves scope (though again, this is error prone and worth documenting)

Reply:
regarding this code 

```c++
auto& st = it->second;
```

This approach is used in many places in the SI code.
Inside the `ss::visist` we're actually writing to the referenced value, I think that we should use some extra synchronization that will prevent container modifications. It's safe for now but we will introduce removals from the archival metadata snapshot and as result we will have to remove data from `_segments`.  But probably, outside of this PR.